File: 2023/04/java/Main.java
package kival.aoc.year2023.day04;

import java.util.ArrayList;
import java.util.List;

import kival.aoc.core.*;

public class Main {
   public final boolean HAS_IO = false;
   public final boolean HAS_ALTERNATE = false;

   public String part1(String input,boolean isTest) {
      int res = 0;
      for (String line : input.split("\n")) {
         List<List<Integer>> sides = new ArrayList<List<Integer>>();
         for (String scratch : line.substring(line.indexOf(":") + 1).split("\\|")) {
            List<Integer> l = new ArrayList<Integer>();
            for (String num : scratch.split(" ")) {
               if (num == "") continue;
               l.add(Integer.parseInt(num));
            }
            sides.add(l);
         }
         int pow = -1;
         for (Integer n : sides.get(0)) if (sides.get(1).contains(n)) pow++;
         res += pow != -1 ? (int)Math.pow(2, pow) : 0;
      }
      return String.valueOf(res);
   }

   public String part2(String input,boolean isTest) {
      String[] lines = input.split("\n");
      int[] instances = new int[lines.length];
      for (int i = 0; i < instances.length; i++)
         instances[i] = 1;
      
      int idx = 0;
      for (String line : lines) {
         List<List<Integer>> sides = new ArrayList<List<Integer>>();
         for (String scratch : line.substring(line.indexOf(":") + 1).split("\\|")) {
            List<Integer> l = new ArrayList<Integer>();
            for (String num : scratch.split(" ")) {
               if (num == "") continue;
               l.add(Integer.parseInt(num));
            }
            sides.add(l);
         }
         int i = 0;
         for (Integer n : sides.get(0)) if (sides.get(1).contains(n)) i++;
         while (i > 0) {
            instances[idx + i] += instances[idx];
            i--;
         }
         idx++;
      }

      int res = 0;
      for (int i = 0; i < instances.length; i++)
         res += instances[i];
      return String.valueOf(res);
   }

   public void main(String[] args) {
      Run.execute(args, this::part1, this::part2, HAS_ALTERNATE, HAS_IO);
   }
}

File: 2023/04/ocaml/main.ml
let options = Hashtbl.create 128;;

Hashtbl.add options "has_alternate" false;;
Hashtbl.add options "has_io" false

let ( +/ ) = Int64.add
let ( -/ ) = Int64.sub
let ( */ ) = Int64.mul
let rec int64_pow x y = if y = 0L then 1L else x */ int64_pow x (y -/ 1L)

let part_1 (input : string) (_is_test : bool) : string =
  input
  |> String.split_on_char '\n'
  |> List.map (fun x ->
    let nums =
      x
      |> String.split_on_char ':'
      |> List.tl
      |> List.hd
      |> String.split_on_char '|'
      |> List.map (fun s ->
        s
        |> String.split_on_char ' '
        |> List.filter (fun s -> String.length s > 0)
        |> List.map int_of_string
        |> Array.of_list)
      |> Array.of_list
    in
    let rec loop (i : int) (j : int64) : int64 =
      if i >= Array.length nums.(0)
      then j
      else if Array.mem nums.(0).(i) nums.(1)
      then loop (i + 1) (j +/ 1L)
      else loop (i + 1) j
    in
    let res = loop 0 (-1L) in
    if res = -1L then 0L else int64_pow 2L res)
  |> List.fold_left ( +/ ) 0L
  |> Int64.to_string
;;

let part_2 (input : string) (_is_test : bool) : string =
  let lines = input |> String.split_on_char '\n' in
  let instances = Array.make (List.length lines) 1L in
  lines
  |> List.mapi (fun i x ->
    let nums =
      x
      |> String.split_on_char ':'
      |> List.tl
      |> List.hd
      |> String.split_on_char '|'
      |> List.map (fun s ->
        s
        |> String.split_on_char ' '
        |> List.filter (fun s -> String.length s > 0)
        |> List.map int_of_string
        |> Array.of_list)
      |> Array.of_list
    in
    let rec loop (i : int) (j : int) : int =
      if i >= Array.length nums.(0)
      then j
      else if Array.mem nums.(0).(i) nums.(1)
      then loop (i + 1) (j + 1)
      else loop (i + 1) j
    in
    let rec aggregate (j : int) : int =
      if j > 0
      then (
        instances.(i + j) <- instances.(i + j) +/ instances.(i);
        aggregate (j - 1))
      else j
    in
    aggregate (loop 0 0))
  |> ignore;
  instances |> Array.fold_left ( +/ ) 0L |> Int64.to_string
;;

let () = Aoc.run Sys.argv part_1 part_2 options
